### PR TITLE
Fusion

### PR DESCRIPTION
**Pull request by DoomDuck**
Integrating every thing together

### CODE REVIEW

**Review by luckylu91 — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -1,44 +1,82 @@
...
+		"front_build": "webpack",
+		"front_build:dev": "webpack --watch",
+		"nest_build": "nest build --path tsconfig.back.json",
+		"nest_start": "nest start --path tsconfig.back.json",
+		"nest_start:dev": "nest start --path tsconfig.back.json --watch",
+		"nest_start:debug": "nest start --path tsconfig.back.json --debug --watch",
+		"nest_start:prod": "node build/back/main.js",
+		"_pongserv_build": "tsc -p src/pong/server/tsconfig.json",
+		"_pongserv_build:dev": "tsc -p src/pong/server/tsconfig.json -w",
+		"_pongserv_start": "node build/pong-temporary-server/server/server.js"
```
Comment (DoomDuck) [inline package.json:18]: I think those scripts are not needed anymore

File: src/pong/server/server.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import express from 'express'
```
Comment (DoomDuck) [inline src/pong/server/server.ts:1]: I think this file in not needed anymore

File: src/pong/common/PhysicBall.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import { Vector3 } from 'three';
```
Comment (DoomDuck) [inline src/pong/common/PhysicBall.ts:1]: We should not have a PhysicalBall in the PR

**Review by DoomDuck — COMMENTED** [review]
Refactoring is needed on game simulation and render

File: src/back/main.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { NestFactory } from '@nestjs/core';
+import { AppModule } from './app.module';
+import { type NestExpressApplication } from '@nestjs/platform-express';
+import { join } from 'path';
+import { INestApplication } from '@nestjs/common';
+
+async function bootstrap() {
+  const app = await NestFactory.create<NestExpressApplication>(AppModule);
+  app.useStaticAssets(join(__dirname, '../..', 'public'), {prefix: "/public"});
+  // console.log(join(__dirname, '../..', 'public'));
```
[the remark this answers is not in the document]
Comment (DoomDuck) [reply src/back/main.ts]: This has been resolved

File: public/global.css [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+canvas {
+    /* position: relative;
+    left: 50%;
+    top: 50%;
+    transform: translateX(-50%); */
+    
+    position: absolute;
+    top: 0px;
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply public/global.css]: top is sufficient to ensure superposition of the game-screen and the game-text

File: public/global.css [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    margin: 0px;
+}
+
+canvas {
+    /* position: relative;
+    left: 50%;
+    top: 50%;
+    transform: translateX(-50%); */
+    
+    position: absolute;
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply public/global.css]: the position is still relative to the containing block

File: public/global.css [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+.game-text {
+    position: absolute;
+    top: 0px;
+}
+
+.player-score {
+    color: #FFF;
+    font-family: sans-serif;
+    background: rgba(0, 0, 0, 0);
+}
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply public/global.css:33]: noted, will be fixed

File: public/img/generate.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply public/img/generate.py:1]: noted, will be fixed

File: src/back/.gitignore [hunk 9]
```diff
@@ -0,0 +1 @@
+/node_modules
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply src/back/.gitignore:1]: noted, will be fixed

File: src/back/.eslintrc.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply src/back/.eslintrc.js:1]: noted, will be fixed

File: src/back/.prettierrc [hunk 11]
```diff
@@ -0,0 +1,4 @@
+{
+  "singleQuote": true,
+  "trailingComma": "all"
+}
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply src/back/.prettierrc:4]: noted, will be fixed

File: tsconfig.common.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
+{
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply tsconfig.common.json:1]: noted, will be fixed

File: tsconfig.back.json [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    ],
+    "outDir": "build/back",
+
+    "importsNotUsedAsValues": "remove",
+    "emitDecoratorMetadata": true,
+    "experimentalDecorators": true,
+    "allowSyntheticDefaultImports": true,
+
+    // "noImplicitAny": true,
+    // "strictNullChecks": true,
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply tsconfig.back.json:22]: issue created

File: src/pong/common/Game.ts [hunk 14]
```diff
@@ -0,0 +1,84 @@
+import { EventEmitter } from "events";
+import { GameEvent, GSettings, PlayerID } from "."
+import { KeyValue, PLAYER1, PLAYER2} from "./constants";
+import { GameState } from "./GameState";
+
+export abstract class Game extends EventEmitter {
+    lastTime: number;
+    timeAccumulated: number;
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply src/pong/common/Game.ts]: noted, will be fixed

File: src/pong/common/collision.ts [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+
+    distanceToCenter.copy(ball.position);
+    distanceToCenter.sub(bar.position);
+    let signX = Math.sign(distanceToCenter.x);
+    let signY = Math.sign(distanceToCenter.y);
+    let x = Math.abs(distanceToCenter.x) - bar.width / 2;
+    let y = Math.abs(distanceToCenter.y) - bar.height / 2;
+    distanceToCenter.x -= signX * (bar.width / 2)
+    distanceToCenter.y -= signY * (bar.height / 2)
+    if (x > 0 && y < 0) {
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply src/pong/common/collision.ts]: noted, will be fixed

File: src/pong/common/collision.ts [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+        // lateral face
+        distanceToCenter.y = 0;
+        horizontal = true;
+    }
+    else if (x < 0 && y > 0) {
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply src/pong/common/collision.ts]: noted, will be fixed

File: src/pong/server/GameManager.ts [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+
+function delay(duration: number) {
+  return new Promise((resolve) => {
+    setTimeout(resolve, duration);
+  });
+}
+
+function otherPlayer(playerId: PlayerID): PlayerID {
+    return (playerId == PLAYER1) ? PLAYER2 : PLAYER1;
+}
```
[the remark this answers is not in the document]
Comment (luckylu91) [reply src/pong/server/GameManager.ts]: noted, will be fixed

File: package.json (hunk 1 above)
Comment (luckylu91) [reply package.json:18]: noted, will be fixed

**Review by Hgallien — APPROVED** [review]
gg

**Pull request merged**